### PR TITLE
Use dynamic namespace for tagging

### PR DESCRIPTION
**Pull request by julcollas**
Here is a pull request for https://github.com/intelsdi-x/snap-plugin-publisher-kairosdb/issues/15

### CODE REVIEW

**Comment by katarzyna-z** [comment]
Thanks for contribution! I've started making review of your code.

File: publisher/publisher.go [hunk 1]
```diff
@@ -80,6 +81,14 @@ func (pub *publisher) GetConfigPolicy() (*cpolicy.ConfigPolicy, error) {
 	r2.Description = "KairosDB port"
 	config.Add(r2)
 
+	r3, err := cpolicy.NewBoolRule("useDynamic", true)
```
Comment (julcollas) [inline publisher/publisher.go]: I've updated small test with the `useDynamic `parameter. I will try to make for tests cases, but I'm really not comfortable with go...

File: README.md [hunk 2]
```diff
@@ -64,6 +64,7 @@ For details on KairosDB, please refer to [documentation](https://kairosdb.github
 Task manifest configuration is described in [snap's documentation](https://github.com/intelsdi-x/snap/blob/master/docs/TASKS.md). In order to use KairosDB publisher you have to add section "publish" then specify following options:
 - `"host"` - KairosDB host address (ex. `"127.0.0.1"`)
 - `"port"` -  KairosDB REST API port (ex. `"8083"`)
+- `"useDynamic"` -  Use dynamic namespaces for metric and tags (ex. `true`)
```
Comment (julcollas) [inline README.md]: I tried be explain a bit more in the example section (https://github.com/julcollas/snap-plugin-publisher-kairosdb/commit/19b16adb0162e7208362c5963b36d3bcbbe4cfa4).

Comment (katarzyna-z) [inline README.md]: What about "Use dynamic elements of namespace as tags for metric"?

Comment (julcollas) [inline README.md]: You right. Thanks
https://github.com/julcollas/snap-plugin-publisher-kairosdb/commit/0ef26626f68119ffc2467200b2be7da7cb923e87

File: README.md [hunk 3]
```diff
@@ -66,10 +66,30 @@ Task manifest configuration is described in [snap's documentation](https://githu
 - `"port"` -  KairosDB REST API port (ex. `"8083"`)
 - `"useDynamic"` -  Use dynamic namespaces for metric and tags (ex. `true`)
 
+To know more about Dynamic Metric Namespace, please refer to [snap's documentation](https://github.com/intelsdi-x/snap/blob/master/docs/METRICS.md#dynamic-metrics)
+
 See example task manifest in [examples/tasks/] (https://github.com/intelsdi-x/snap-plugin-publisher-kairosdb/blob/master/examples/tasks/).
 
 ### Examples
-Example of running [psutil collector plugin](https://github.com/intelsdi-x/snap-plugin-collector-psutil) and publishing data to KairosDB.
+Here are samples to illustrate the differences for `useDynamic` flag with [plugin-collector-cpu](https://github.com/intelsdi-x/snap-plugin-collector-cpu)
```
Comment (katarzyna-z) [inline README.md]: Could you make this example with [snap-plugin-collector-psutil](https://github.com/intelsdi-x/snap-plugin-collector-psutil)? We will have a consistent example. snap-plugin-collector-psutil has also dynamic metrics and it is easy to use as snap-plugin-collector-cpu.

Comment (julcollas) [inline README.md]: done

File: publisher/publisher.go (hunk 1 above)
Comment (katarzyna-z) [inline publisher/publisher.go]: As you are not comfortable with go, I've prepared something. You can replace existing code of TestPublish with the code which I prepared, please see:
```golang
func (s *KairosSuite) TestPublish() {
	Convey("Given snap KairosDB publisher testing", s.T(), func() {
		metrics := []plugin.MetricType{
			plugin.MetricType{
				Namespace_: core.NewNamespace("foo"),
				Timestamp_: time.Now(),
				Tags_:      map[string]string{"hostname": "127.0.0.1"},
				Data_:      43,
			},
			plugin.MetricType{
				Namespace_: core.NewNamespace("bar").AddDynamicElement("dynamic_element", "dynamic element of namespace"),
				Timestamp_: time.Now(),
				Tags_:      map[string]string{"hostname": "127.0.0.1"},
				Data_:      44,
			},
		}

		Convey("Publishing dynamic elements of namespace as tags for metric", func() {
			var buf bytes.Buffer
			config := make(map[string]ctypes.ConfigValue)
			u, _ := url.Parse(s.server.URL)
			hostPort := strings.Split(u.Host, ":")
			config["host"] = ctypes.ConfigValueStr{Value: hostPort[0]}
			port, _ := strconv.Atoi(hostPort[1])
			config["port"] = ctypes.ConfigValueInt{Value: port}
			config["useDynamic"] = ctypes.ConfigValueBool{Value: true}

			kp := New()
			cp, _ := kp.GetConfigPolicy()
			cfg, _ := cp.Get([]string{""}).Process(config)

			Convey("Publish provided metrics", func() {

				buf.Reset()
				enc := gob.NewEncoder(&buf)
				enc.Encode(metrics)
				err := kp.Publish(plugin.SnapGOBContentType, buf.Bytes(), *cfg)
				So(err, ShouldBeNil)
			})
		})

		Convey("Publishing without dynamic elements of namespace as tags for metric", func() {
			var buf bytes.Buffer
			config := make(map[string]ctypes.ConfigValue)
			u, _ := url.Parse(s.server.URL)
			hostPort := strings.Split(u.Host, ":")
			config["host"] = ctypes.ConfigValueStr{Value: hostPort[0]}
			port, _ := strconv.Atoi(hostPort[1])
			config["port"] = ctypes.ConfigValueInt{Value: port}
			config["useDynamic"] = ctypes.ConfigValueBool{Value: false}

			kp := New()
			cp, _ := kp.GetConfigPolicy()
			cfg, _ := cp.Get([]string{""}).Process(config)

			Convey("Publish provided metrics", func() {

				buf.Reset()
				enc := gob.NewEncoder(&buf)
				enc.Encode(metrics)
				err := kp.Publish(plugin.SnapGOBContentType, buf.Bytes(), *cfg)
				So(err, ShouldBeNil)
			})
		})
	})
}
```
You can replace existing code of `TestPublish` with the code which I prepared.

Comment (julcollas) [inline publisher/publisher.go]: I confirm, I'm not confident enough for the moment with go to write that :)

**Comment by katarzyna-z** [comment]
@julcollas I tested this pull request and it works as expected so please squash commits into single commit.

**Comment by julcollas** [comment]
@katarzyna-z , can I squash from my side ?

**Comment by katarzyna-z** [comment]
@julcollas yes, you can, see for example [this](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/) instruction.

**Comment by katarzyna-z** [comment]
@julcollas I've tried to help you with squashing of commits but unfortunately this pr has been closed. Are you able to open a new pr?

**Comment by julcollas** [comment]
@katarzyna-z , I did 'not squash it correctly ?

**Comment by katarzyna-z** [comment]
Unfortunately, yes. I'll try to help you.

**Comment by katarzyna-z** [comment]
@julcollas Could you give me write access to you fork? I'll prepare a correct branch. You can make it on your fork in "Settings" > "Collaborators"

**Comment by julcollas** [comment]
@katarzyna-z , thansk. You should now have access.

**Comment by katarzyna-z** [comment]
@julcollas I've created a new [branch](https://github.com/julcollas/snap-plugin-publisher-kairosdb/tree/pullrequest16) and I've opened a new [pull request](https://github.com/intelsdi-x/snap-plugin-publisher-kairosdb/pull/17).
Once again, thanks for your contribution! :rose:

**Comment by julcollas** [comment]
Thanks a lot for you help.
Next time I will be better in go and git :)

**Pull request closed without merging**